* Enzymatic Methylation Sequencing Bioinformatics Processes
** Repository administration
*** Configuration
#+begin_src yaml :tangle ./config/emseq.yaml
conda-env: emseq

sourced-conda-envs:
  - name: biotools
    version: 1.0.1
  - name: basecamp
    version: 1.1.0

#+end_src
*** Conda
**** Environmental YAMLs
#+begin_src yaml :tangle ./config/emseq-conda-env.yaml
name: emseq
channels:
  - conda-forge
  - bioconda

dependencies:
  - biscuit=1.6.0.20241216=*_1
  - dupsifter=1.3.0.20241113=*_1
  - bioconductor-methylkit
  - r-data.table
#+end_src
**** Testing
#+begin_src bash
cd ~/repos/emseq

ENV_NAME="emseq"

# Create the environment
conda create --name "$ENV_NAME" -y

mapfile -t CONDA_ENVS < <(yqgo e -N '.sourced-conda-envs[] | .name + ":" + .version' ./config/emseq.yaml)

for entry in "${CONDA_ENVS[@]}"; do
  name="${entry%%:*}"
  version="${entry##*:}"
  url="https://raw.githubusercontent.com/jeszyman/${name}/v${version}/${name}_env.yaml"
  dest="/tmp/${name}_env.yaml"
  curl -fsSL -o "$dest" "$url" \
    && mamba env update --name "$ENV_NAME" --file "$dest"
done

mamba env update --name "$ENV_NAME" --file ./config/emseq-conda-env.yaml
#+end_src

*** Snakemake
*** README
Directory Structure
#+begin_example
-- Main
   |-- config
   |-- resources
   |-- results
   |-- scripts
   |-- test
   |-- tools
   |-- workflows
#+end_example

**** Prerequisites
**** Change Log
** Processes
*** Methylation sequence processing
- Consider reference w/ decoys https://chatgpt.com/c/67c1c299-f8ec-8005-a2ba-59e05af12369
******** Biscuit index
https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.15_GRCh38/seqs_for_alignment_pipelines.ucsc_ids/GCA_000001405.15_GRCh38_no_alt_plus_hs38d1_analysis_set.fna.bwa_index.tar.gz


#+begin_src bash
aria2c -c -x 10 -s 10 -m 5 -d $data_dir/inputs/ \
    -o Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz \
    https://ftp.ensembl.org/pub/release-113/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz

cd $data_dir/ref

mkdir $data_dir/ref/biscuit

# unzip the fasta.gz to ref
gunzip -c $data_dir/inputs/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz > $data_dir/ref/biscuit/Homo_sapiens.GRCh38.dna.primary_assembly.fa

samtools faidx $data_dir/ref/biscuit/Homo_sapiens.GRCh38.dna.primary_assembly.fa

nohup biscuit index $data_dir/ref/biscuit/Homo_sapiens.GRCh38.dna.primary_assembly.fa & disown

#+end_src

** Ideas
